src/fabric.c

```diff
@@ -568,59 +568,77 @@ static void ofi_set_prov_attr(struct fi_fabric_attr *attr,
  * 2. If a utility provider is specified, return it over any* core provider.
  * 3. If a core provider is specified, return any utility provider that can
  *    layer over it, plus the core provider itself, if possible.
- * 4* A utility provider will not layer over the sockets provider unless the
+ * 4. A utility provider will not layer over the sockets provider unless the
  *    user explicitly requests that combination.
+ * 5. If both the providers are utility providers or if more than two providers
+ *    are specified, the rightmost provider would be compared.
+ * 6. If any provider has a caret symbol "^" is prefixed before any provider
+ *    name it would be excluded (internal use only). These excluded providers
+ *    should be listed only at the end.
  *
  * Utility providers use an internal flag, OFI_CORE_PROV_ONLY, to indicate
  * that only core providers should respond to an fi_getinfo query.  This
  * prevents utility providers from layering over other utility providers.
  */
 static int ofi_layering_ok(const struct fi_provider *provider,
-			   const char *util_name, size_t util_len,
-			   const char *core_name, size_t core_len,
+			   char **prov_vec, size_t count,
 			   uint64_t flags)
 {
+	char *prov_name;
+	int i;
+
+	/* Excluded providers are assumed to be at the end */
+	for (i = count - 1; i >= 0; i--) {
+		if (prov_vec[i][0] != '^')
+		    break;
+
+		if (!strcasecmp(&prov_vec[i][1], provider->name))
+			return 0;
+	}
+	count = i + 1;
+
 	if (flags & OFI_CORE_PROV_ONLY) {
+		assert((count == 1) || (count == 0));
 		if (ofi_is_util_prov(provider)) {
 			FI_INFO(&core_prov, FI_LOG_CORE,
 				"Need core provider, skipping util %s\n",
 				provider->name);
 			return 0;
 		}
 
-		if ((!core_len || !core_name) &&
-		    !strcasecmp(provider->name, "sockets")) {
+		if ((count == 0) && !strcasecmp(provider->name, "sockets")) {
 			FI_INFO(&core_prov, FI_LOG_CORE,
 				"Skipping util;sockets layering\n");
 			return 0;
 		}
 	}
 
-	if (util_len && util_name) {
-		assert(!(flags & OFI_CORE_PROV_ONLY));
-		if ((strlen(provider->name) != util_len) ||
-		    strncasecmp(util_name, provider->name, util_len))
-			return 0;
+	if (!count)
+		return 1;
 
-	} else if (core_len && core_name) {
-		if (!strncasecmp(core_name, "sockets", core_len) &&
-		    ofi_is_util_prov(provider)) {
+	if ((count == 1) && ofi_is_util_prov(provider) &&
+	    !ofi_has_util_prefix(prov_vec[0])) {
+		if (!strcasecmp(prov_vec[0], "sockets")) {
 			FI_INFO(&core_prov, FI_LOG_CORE,
 				"Sockets requested, skipping util layering\n");
 			return 0;
+		} else {
+			return 1;
 		}
-
-		if (!ofi_is_util_prov(provider) &&
-		    ((strlen(provider->name) != core_len) ||
-		     strncasecmp(core_name, provider->name, core_len)))
-			return 0;
 	}
 
-	return 1;
+	if ((count == 2) && ofi_has_util_prefix(prov_vec[0]) &&
+	    !ofi_has_util_prefix(prov_vec[1]))
+		prov_name = prov_vec[0];
+	else
+		prov_name = prov_vec[count - 1];
```

Comment:
The higher-level logic for the above series of if checks doesn't clearly convey what the rules are to include/exclude a provider based on the remaining provider list.  Specifically, it's not clear why the count value should matter when determining if the layering is okay.   At the very least we probably need to add comments (yikes) here to describe the rules.